### DISCUSSION TITLE
Keep errors while submiting to the server

### DISCUSSION THREAD

**Discussion by andresgutgon** [IdoPesok/zsa#143] [General]
# What?
Hi, I'm using `useServerAction` to save some data into my server. It's working fine but I see a glitch while `isPending`

The errors in the server are reset and the UI jumbs.

This is my form:
``` javascript
  const { data, executeFormAction, isPending, error, ...rest } = useServerAction(
    onboarding,
    {
      initialData: { name, username },
    },
  )
  const errorData = error?.data
  const fieldErrors = error?.fieldErrors
  return (
      <Input
        label='Your username'
        type='text'
        name='username'
        errors={fieldErrors?.username}
        defaultValue={data?.username ?? errorData?.username}
        onChange={(e) => setUserName(e.target.value)}
        placeholder='Pick a username'
        description={<UsernameDescription username={currentUsername} />}
      />
  )
```

This is my action:
``` javascript
export const onboarding = authProcedure
  .createServerAction()
  .input(input, { type: 'formData' })
  .output(input)
  .experimental_shapeError(({ err, typedData }) => {
    if (err instanceof ZSAError && err.code !== 'INPUT_PARSE_ERROR') {
      return {
        code: err.code,
        message: err.message,
      }
    }

    return {
      data: typedData?.inputRaw,
      fieldErrors: typedData?.inputParseErrors?.formattedErrors,
    }
  })
  .handler(async ({ input, ctx: { user } }) => {
    await validateUniqueUsername({ data: input, user })

    const result = await finishOnboarding({ user, data: input })
    const value = result.unwrap()
    await updateSession({
      user: { name: value.name, username: value.username },
    })

    redirect('/')
  })
```

This is my UI. Check the disappearing red errors when submitting

https://github.com/IdoPesok/zsa/assets/49499/b17d4941-0091-4758-a184-23fd6c254177

**Comment by andresgutgon** [comment]
Am I using it ok? Is there a way of avoiding this glitch?

**Comment by andresgutgon** [comment]
Also, to keep previous data, I had to use `shapeError` and get `typedData?.inputRaw`. How hard would always be to return data already present in the form? 

I say this because otherwise, all inputs are empty when one of them have a validation error no?

**Comment by IdoPesok** [comment]
Hi, thanks for bringing this up!

I understand the issue. Going to compare with other libraries and report back shortly with next steps.

In the meantime -- if you want to try -- you should be able to get this persistence with [useActionState](https://zsa.vercel.app/docs/forms#useactionstate). This also supports progressive enhancement.

> always be to return data already present in the form?

Do you mean to return `inputRaw` in a default TZSAError without needing to do shape error?

**Comment by andresgutgon** [comment]
> Do you mean to return inputRaw in a default TZSAError without needing to do shape error?

Yes!

**Comment by andresgutgon** [comment]
Using `state` gives me this TS error

I'm in latest `zsa`. In fact I'm debugging it locally with `npm link`
<img width="1245" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/62ce1c95-b10a-4adc-bc31-995d03b93ce8">

**Comment by IdoPesok** [comment]
What version of React is this?

**Comment by IdoPesok** [comment]
Also if you are using npm link, do you want to give the input schema functions a shot? Going to merge soon would be awesome if you can test that branch in your project and give feedback.

**Comment by andresgutgon** [comment]
> What version of React is this?

I'm living on the edge xD

``` json
{
    "next": "14.3.0-canary.59",
    "next-auth": "5.0.0-beta.19",
    "react": "19.0.0-beta-26f2496093-20240514",
    "react-dom": "19.0.0-beta-26f2496093-20240514"
}
```

Maybe is because of types?
``` json
    "@types/react": "^18.2.61",
    "@types/react-dom": "^18.2.19",
```

**Comment by andresgutgon** [comment]
> Also if you are using npm link, do you want to give the input schema functions a shot?

I can try that. But would be ideal to use `useServerAction`

And out of the box get:

1. `data` persisted between attempts
2. Do not erase errors until the response come from the server (action)

But I think that would be ideal without needing to customize errors or other parts of the lib. It would make a straightforward understand of the model

**Comment by IdoPesok** [comment]
Ah yes, I meant outside of this issue and related to #120 so you don't need to do validate unique username in the handler.

**Comment by andresgutgon** [comment]
Ah!! yes yes I didn't understand you. Let me check that

**Comment by IdoPesok** [comment]
Hi, the formatted errors come from [Zod formatted errors](https://zod.dev/ERROR_HANDLING?id=formatting-errors). Same thing with field errors, and form errors. These are standard Zod error schemas that people use. We can be more clear in our docs to explain this.

**Comment by andresgutgon** [comment]
~~I see, but my question is. For having formatted errors do I need to use `shapeError`. I'll try with stable branch but i think it's working without that method~~

It does indeed. And also in your PR. All was my fault https://github.com/IdoPesok/zsa/issues/121#issuecomment-2171686993

**Comment by andresgutgon** [comment]
Playing with this https://github.com/IdoPesok/zsa/pull/120 is failing for me

Basically I'm not sure what I have to pass in the `input`
I'm passing a `z.object` but is failing
```javascript
input(async ({ ctx }) => z.object(...), { type: 'formData'})
```
<img width="605" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/82bb2f61-47e3-4204-979c-35fc0e0c46dd">


I see this in Chrome Dev tools
<img width="360" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/8b2382a2-3740-4aa5-a666-1261e09c8e5a">

I think that PR could have a regression

The same code in my app gets the error on `error` key from `useServerAction` hook
<img width="521" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/7f6263c6-aee0-48e2-920a-d74d8682de17">
<img width="461" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/b8601431-3c1f-465c-9de3-e14bf0bf0a77">


But with that PR I get the error about my `ZodObject` not being an instance of `z.ZodType`. No idea what could be wrong. My ZodObject is indeed an instance of zodType

**Comment by andresgutgon** [comment]
Doing these 2 changes it works. No need for custom `shapeError`

Looks like something related with `instanceof` but no idea why these 2 return false

<img width="980" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/6438dd5e-d9a1-44fa-adb7-3294f98af2d6">

**Comment by andresgutgon** [comment]
Maybe is related https://github.com/colinhacks/zod/issues/2241 or this one https://github.com/colinhacks/zod/issues/2746

**Comment by andresgutgon** [comment]
**ALL_MY_FAULT** 😂 

Oh man, NodeJS is hard. 
What happened is your monorepo use `NPM` and I was linking from my monorepo with `pnpm link ~/code/opensource/zsa/packages/zsa` and it was taking Zod from pnpm and so the instance was different when checking `instanceof` 

What I did instead is 

```json
{
  "zsa": "file:../../../../opensource/zsa/packages/zsa",
  "zsa-react": "file:../../../../opensource/zsa/packages/zsa-react",
}
```

Fuck, this was a good lesson. 

Your PR works great `input(function)` does what it promise 🎉

<img width="458" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/60512139-e2fc-419a-8ab1-45bb274e3363">



I'll move now to investigate the problems in this issue

OFF-TOPIC:
What's the use case for `retry`? Reading the code it makes everything more complicated. I'm curious to know what made you implement retries in this package. I can't imagine any good cases

**Comment by IdoPesok** [comment]
Sorry for late response, was traveling this wknd!

> Oh man, NodeJS is hard. What happened is your monorepo use `NPM` and I was linking from my monorepo with `pnpm link ~/code/opensource/zsa/packages/zsa` and it was taking Zod from pnpm and so the instance was different when checking `instanceof`

Ah this is super tricky! I was so confused why the instance of wasn't getting the ZodObject in your first message. Good job debugging this.

> Your PR works great input(function) does what it promise 

Awesome! Happy to hear! Still writing some more tests. Excited to merge it.

> What's the use case for retry? 

Sometimes if there are rate limits in effect it could be useful to retry. This is also available in libraries such as `react-query`. I agree it does make things more complicated 😄 I had a branch with many more features such as polling and caching but figured people just should use `react-query` with `zsa` since its already so good.

Okay, I am going to move this conversation to #124 as the code is there.

**Comment by andresgutgon** [comment]
> Polling and caching

Yep, these are frontend concerns and because zsa already integrates with react-query there's no need to bring that complexity to this code.

> Okay, I am going to move this conversation to https://github.com/IdoPesok/zsa/pull/124 as the code is there.

Nice, happy to change anything
